app/javascript/mastodon/components/poll.js

```diff
@@ -32,8 +32,36 @@ class Poll extends ImmutablePureComponent {
 
   state = {
     selected: {},
+    expired: this.props.poll.get('expired') || (new Date(this.props.poll.get('expires_at'))).getTime() < this.props.intl.now(),
```

Comment:
Set this to null.

Create `_setupTimer` method. Call this method from `componentDidMount`. Call this method from `componentWillReceiveProps` if `nextProps.poll.get('expired') !== this.props.poll.get('expired')` etc.

Reply:
Setting it to null and then changing it in `componentDidMount` would incur an useless extra render cycle for every poll.

Reply:
Well, realistically, there's no reason to store `expired` in state, for example `RelativeTimestamp` component stores *current timestamp* in the state, and then the render method takes care of calculating the difference. That's probably the cleanest approach.

Reply:
I'd say there's no reason to store the current timestamp either, tbh.
Changed it to not use the deprecated `componentWillReceiveProps` and refactor it, it's cleaner imho, and should not do useless renders. However, it will re-fire the timer each time a poll gets updated even if its expiration date does not change. This is infrequent and cheaper than a render, though.